fix(ux): show report button on too many writes error

nextchamp-saqib opened the pull request:
Show report button on too many writes error. This will help to pinpoint the frappe app which causes the error. 

`no-docs`

nextchamp-saqib committed:
feat: log errors on too many writes to ease debugging

ankush commented:
Can we somehow also find which app did it?

Most SQL writes will have traceback that follows this structure:

1. Frappe framework initiates something (like submitting a doc)
2. erpnext/third-party app / server script hooks into process and calls ORM method
3. orm method (e.g. SQL) makes a write that finally triggers failure. 

We need to somehow find 2nd only :P I think it shouldn't be too hard if we traverse the stack in reverse order and ignore database files.


PS: I'll try to push something for this; if doable. 

Edit: Done

<img width="1440" alt="Screenshot 2022-01-17 at 6 20 57 PM" src="https://user-images.githubusercontent.com/9079960/149772447-63bab8df-c681-4573-ad48-e30e04c716d0.png">

ankush committed:
fix(ux): extract and show which app/file/function caused too many writes

ankush committed:
fix: postgres error checking assumes exception is postgres exception

ankush committed:
fix: postgres error checking assumes exception is postgres exception

ankush commented on frappe/utils/error.py:
```diff
@@ -233,3 +233,34 @@ def wrapper_raise_error_on_no_output(*args, **kwargs):
 		return wrapper_raise_error_on_no_output
 	return decorator_raise_error_on_no_output
+
+def get_app_details_from_stack(skip_frames=0, ignore_files=()):
```
```suggestion
def get_app_details_from_stack(skip_frames=1, ignore_files=()):
```

Maybe 1 is better default here. 

_Typically_ stack is like:  some_app -> your_function -> this function. So last 2 frames need to be skipped.

ankush committed:
fix: better default for skipping frames

Typically stack is like: some_app -> your_function -> this function. So last 2 frames need to be skipped.

ankush committed:
fix: better default for skipping frames

Typically stack is like: some_app -> your_function -> this function. So last 2 frames need to be skipped.

gavindsouza commented:
The error message still seems off. I think we can do better. I'll come up with something clear-er; if anyone has something more concise, clear and easy to understand drop a suggestion, thanks!

rmehta commented:
@nextchamp-saqib server tests are failing?

nextchamp-saqib committed:
Merge branch 'develop' of https://github.com/frappe/frappe into log-error-on-over-commits

ankush commented:
Test started failing after changing default because test runner stack trace is much shorter than usual frappe ORM actions so it skipped right past all frappe files to stdlib code. 🥲  Anyway, this shouldn't affect other usual code as demonstrated in other tests in `test_db` which is passing :p 


edit: fixed.

ankush commented:
> The error message still seems off. I think we can do better. I'll come up with something clear-er; if anyone has something more concise, clear and easy to understand drop a suggestion, thanks!


@gavindsouza any update on this? Two of us already tried to come up with best message possible, but the reality is there is almost nothing a regular user can do to recover from this error. A dev can still get a useful hint about what caused too many writes though. 



Edit: maybe this should show "report" button so traceback can be immediately submitted instead of storing in logs @nextchamp-saqib

ankush commented:
This is going nowhere, IMO let's just throw `Exception` instead of `frappe.ValidationError` that will show the "report" button and also enable sharing of full traceback. 

@nextchamp-saqib @gavindsouza

nextchamp-saqib committed:
Merge branch 'develop' of https://github.com/frappe/frappe into log-error-on-over-commits

nextchamp-saqib committed:
Merge branch 'log-error-on-over-commits' of https://github.com/nextchamp-saqib/frappe into log-error-on-over-commits

nextchamp-saqib committed:
refactor: show Report button instead of logging

ankush committed:
revert: unnecessary new functionality

ankush committed:
test: assert exact exception

ankush commented on frappe/database/postgres/database.py:
```diff
@@ -170,11 +170,11 @@ def is_duplicate_entry(e):
 	@staticmethod
 	def is_primary_key_violation(e):
-		return e.pgcode == '23505' and '_pkey' in cstr(e.args[0])
+		return getattr(e, "pgcode", None) == '23505' and '_pkey' in cstr(e.args[0])
```
Leaving these changes in as these have to be fixed someday 😄

ankush commented:
@mergifyio backport version-13-hotifx

ankush commented:
@Mergifyio backport version-13-hotfix
